include/sys/dmu.h

```diff
@@ -797,9 +800,11 @@ int dmu_write_uio_dnode(dnode_t *dn, struct uio *uio, uint64_t size,
 #endif
 struct arc_buf *dmu_request_arcbuf(dmu_buf_t *handle, int size);
 void dmu_return_arcbuf(struct arc_buf *buf);
-void dmu_assign_arcbuf(dmu_buf_t *handle, uint64_t offset, struct arc_buf *buf,
-    dmu_tx_t *tx);
-void dmu_assign_arcbuf_impl(dmu_buf_t *handle, struct arc_buf *buf,
+void dmu_assign_arcbuf(objset_t *os, uint64_t object, uint64_t offset,
+    struct arc_buf *buf, dmu_tx_t *tx);
+void dmu_assign_arcbuf_by_dbuf(dmu_buf_t *handle, uint64_t offset,
+    struct arc_buf *buf, dmu_tx_t *tx);
+void dmu_assign_arcbuf_by_dbuf_impl(dmu_buf_t *handle, struct arc_buf *buf,
```

Comment:
This is a good time to remove this `_impl` function from both the header and it implementations from `module/zfs/dmu.c`.  The single call site in `module/zfs/dmu_send.c` can be updated to do the cast as a simplification.